docs/design/019-file-api.md

```diff
@@ -0,0 +1,226 @@
+
+### Context
+Currently, files cannot be opened from within a function executed by a VU, only in the init context.
+
+This is due to k6's design for distributed execution, particularly in the cloud. K6 runs the init context once, gathers resources, including files, and sends them to other instances where VU code runs.
+
+### Assumptions
+- The solution should be native to k6, with the code being in the k6 repository and maintained by the k6 OSS team.
+- The new functionality's API should be intuitive for users familiar with filesystem APIs from other languages and technologies.
+- The module should introduce only essential APIs: read-only operations.
+- The proposed API should not affect file handling within HTTP requests.
+
+### Requirements
+- The solution should allow file interaction during k6 script execution.
+- The solution should resort to asynchronous operations as much as possible.
+- The module should operate seamlessly in local and cloud setups.
+- The module should strive to optimize memory usage when handling files.
+
+### Scope
+
+#### In scope
+- The proposed solution will offer an alternative to the existing `open` function.
+- The solution will provide a File concept that supports read-only operations: reading file chunks into a buffer, reading the entire file content, and file seeking.
+- The proposal could suggest changes to improve overall memory usage in k6 scripts.
+
+#### Out of scope
+- This proposal doesn't aim to solve memory consumption issues when using file content in HTTP requests, although minor improvements might be suggested.
+- Changes to how k6 gathers resources, particularly files, are out of scope (ideas are proposed in the "Future Improvements" section).
+- Despite the proposed API's potential to support them, write operations are beyond this proposal's scope.
+
+## Proposed solution
+
+We suggest implementing a minimalist, experimental file system (`fs`) module based on [Deno's fs module](https://deno.land/api@v1.32.3?s=Deno.open). The new module will allow users to interact with files, separating text and binary files. The module will provide an `open` function that returns a file handle/view for performing read operations.
+
+The initial API will mostly be asynchronous, except for the `open` functionality which will be synchronous due to the current lack of support for `await` operations within the init context.
+
+The API will have the following characteristics:
+- Load file content exactly once into memory.
+- Provide each VU a copy of the file handle pointing to a unique memory area to avoid copying the whole buffer for each VU.
+- Ensure each file handle has unique offsets allowing each VU to work independently.
```

Comment:
nit: This is worded in a way that sounds as if a single VU can not open the *same* file multiple times and have differnet offsets - which isn't true from the code that I looked at.